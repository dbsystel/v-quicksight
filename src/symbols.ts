export const EmbeddingContextInjectionKey = Symbol(
  "The key to be used with inject when accessing the created embeddingContext for usage with the amazon-quicksight-embedding-sdk",
)
