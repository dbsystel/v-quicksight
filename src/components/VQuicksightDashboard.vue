<script setup lang="ts">
import type { VQuicksightDashboardContentOptions, VQuicksightFrameOptions } from '../types'
import type {
  DashboardContentOptions,
  DashboardFrame,
  EmbeddingContext,
  ExperienceFrameMetadata,
  FrameOptions,
  Parameter,
  SimpleChangeEvent,
  SimpleMessageEvent
} from 'amazon-quicksight-embedding-sdk/dist/types'
import { nanoid } from 'nanoid'
import type { Ref } from 'vue'
import { computed, inject, ref, watch } from 'vue'
import { EmbeddingContextInjectionKey } from '../symbols'

const props = withDefaults(
  defineProps<
    VQuicksightFrameOptions &
      VQuicksightDashboardContentOptions & {
        container?: string | HTMLElement // override from FrameOptions and make it optional
        dashboard?: string
        id?: string
      }
  >(),
  {
    width: '100%',
    height: '100%',
    resizeHeightOnSizeChangedEvent: true,
    withIframePlaceholder: false,
    parameters: () => [],
    sheetOptions: () => ({}),
    toolbarOptions: () => ({}),
    attributionOptions: () => ({})
  }
)

const emit = defineEmits<{
  (e: 'change', data: { changeEvent: SimpleChangeEvent; metadata?: ExperienceFrameMetadata }): void
  (
    e: 'message',
    data: { messageEvent: SimpleMessageEvent; experienceMetadata?: ExperienceFrameMetadata }
  ): void
}>()

const embeddingContext = inject<Ref<EmbeddingContext>>(EmbeddingContextInjectionKey)

const dashboardFrame = ref<DashboardFrame>()
const changedDashboard = ref<string>()

const containerId = computed(() => props.id || `v-quicksight-dashboard-${nanoid(6)}`)
const frameOptions = computed<FrameOptions>(() => {
  return {
    url: props.url,
    container: props.container || `#${containerId.value}`,
    height: props.height,
    width: props.width,
    className: props.className,
    resizeHeightOnSizeChangedEvent: props.resizeHeightOnSizeChangedEvent,
    withIframePlaceholder: props.withIframePlaceholder,
    onChange: (changeEvent, metadata) => {
      emit('change', { changeEvent, metadata })
    }
  }
})
const contentOptions = computed<DashboardContentOptions>(() => {
  return {
    parameters: props.parameters,
    locale: props.locale,
    sheetOptions: props.sheetOptions,
    toolbarOptions: props.toolbarOptions,
    attributionOptions: props.attributionOptions,
    onMessage: async (messageEvent, experienceMetadata) => {
      emit('message', { messageEvent, experienceMetadata })
    }
  }
})

async function embedDashboard(ctx: EmbeddingContext, dashboard?: string) {
  dashboardFrame.value = await ctx.embedDashboard(frameOptions.value, contentOptions.value)
  if (dashboard) {
    navigateToDashboard(dashboardFrame.value, dashboard)
  }
}

async function navigateToDashboard(frame: DashboardFrame, dashboard: string) {
  await frame.navigateToDashboard(dashboard, {
    parameters: props.parameters
  })
}

async function setParameters(frame: DashboardFrame, parameters: Parameter[]) {
  return await frame.setParameters(parameters)
}

watch(
  () => embeddingContext?.value,
  (newValue, oldValue) => {
    if (newValue && !oldValue) {
      embedDashboard(newValue)
    }
  },
  { immediate: true }
)

watch(
  () => props.dashboard,
  (newValue) => {
    changedDashboard.value = newValue
    if (newValue && dashboardFrame.value) {
      navigateToDashboard(dashboardFrame.value, newValue)
    } else if (embeddingContext?.value) {
      embedDashboard(embeddingContext?.value, newValue)
    } else {
      console.error('could not find any embeddingContext for embedding / updating dashboard')
    }
  }
)

watch(
  () => props.parameters,
  (newValue) => {
    console.log('UPDATE PARAMETERS', {
      newValue,
      dashboardFrame: dashboardFrame.value,
      currentDashboard: changedDashboard.value,
      dashboard: props.dashboard
    })
    if (newValue && dashboardFrame.value) {
      // only set parameters when dashboard has not been changed.
      // Otherwise changes will be handled by navigateToDashboard
      if (!changedDashboard.value || changedDashboard.value === props.dashboard) {
        setParameters(dashboardFrame.value, newValue)
      }
    } else if (embeddingContext?.value) {
      embedDashboard(embeddingContext?.value)
    } else {
      console.error('could not find any embeddingContext for updating parameters')
    }
  },
  { deep: true }
)
</script>

<template>
  <div class="v-quicksight-dashboard" :id="containerId"></div>
</template>
