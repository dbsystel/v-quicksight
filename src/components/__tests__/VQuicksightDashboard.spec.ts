import VQuicksightDashboard from '../VQuicksightDashboard.vue'
import { EmbeddingContextInjectionKey } from '../../symbols'
import { mount } from '@vue/test-utils'
import { describe, expect, it, vi } from 'vitest'
import { nextTick, ref } from 'vue'

const frameOptions = {
  container: '#custom-el',
  width: '300px',
  height: '700px',
  resizeHeightOnSizeChangedEvent: false,
  withIframePlaceholder: true,
  className: 'my-custom-class'
}
const contentOptions = {
  parameters: [
    {
      Name: 'minimum',
      Values: [10]
    }
  ],
  locale: 'en-US',
  sheetOptions: {
    initialSheetId: 'my-sheet-1',
    singleSheet: true,
    emitSizeChangedEventOnSheetChange: true
  },
  toolbarOptions: {
    export: true,
    undoRedo: true,
    reset: true
  },
  attributionOptions: {
    overlayContent: true
  }
}

describe('VQuicksightDashboard', () => {
  function setupComponent(props: Record<string, any> = {}) {
    const dashboardFrame = {
      setParameters: vi.fn().mockResolvedValue({ success: true }),
      navigateToDashboard: vi.fn().mockResolvedValue({ success: true })
    }
    const embeddingContext = {
      embedDashboard: vi.fn().mockResolvedValue(dashboardFrame)
    }

    const component = mount(VQuicksightDashboard, {
      global: {
        provide: {
          [EmbeddingContextInjectionKey]: ref(embeddingContext)
        }
      },
      props: {
        ...props,
        url: 'https://my-embed-url.example.org',
        id: 'v-quicksight-dashboard-1'
      }
    })
    return { component, dashboardFrame, embeddingContext }
  }

  it('renders as expected', () => {
    const { component } = setupComponent()
    expect(component.element).toMatchSnapshot()
  })

  it('should call embedDashboard with default parameters', async () => {
    const { embeddingContext } = setupComponent()
    await nextTick()
    expect(embeddingContext.embedDashboard).toHaveBeenCalledOnce()
    expect(embeddingContext.embedDashboard).toHaveBeenCalledWith(
      {
        className: undefined,
        container: '#v-quicksight-dashboard-1',
        height: '100%',
        onChange: expect.anything(),
        resizeHeightOnSizeChangedEvent: true,
        url: 'https://my-embed-url.example.org',
        width: '100%',
        withIframePlaceholder: false
      },
      {
        attributionOptions: {},
        locale: undefined,
        onMessage: expect.anything(),
        parameters: [],
        sheetOptions: {},
        toolbarOptions: {}
      }
    )
  })

  it('should call embedDashboard with customized parameters', () => {
    const { embeddingContext } = setupComponent({ ...frameOptions, ...contentOptions })
    expect(embeddingContext.embedDashboard).toHaveBeenCalledOnce()
    expect(embeddingContext.embedDashboard).toHaveBeenCalledWith(
      {
        ...frameOptions,
        onChange: expect.anything(),
        url: 'https://my-embed-url.example.org'
      },
      {
        ...contentOptions,
        onMessage: expect.anything()
      }
    )
  })

  it('should call navigateToDashboard when the passed dashboard name changes', async () => {
    const { dashboardFrame, component } = setupComponent()
    await nextTick() // wait until first setup is finished
    await component.setProps({ dashboard: 'another-dashboard' })
    expect(dashboardFrame.navigateToDashboard).toHaveBeenCalledOnce()
    expect(dashboardFrame.navigateToDashboard).toHaveBeenCalledWith('another-dashboard', {
      parameters: []
    })
  })

  it('should call setParameters when the passed parameters are changing', async () => {
    const { dashboardFrame, component } = setupComponent()
    await nextTick() // wait until first setup is finished
    await component.setProps({ parameters: contentOptions.parameters })
    expect(dashboardFrame.setParameters).toHaveBeenCalledOnce()
    expect(dashboardFrame.setParameters).toHaveBeenCalledWith(contentOptions.parameters)
  })

  it('should call navigateToDashboard when the passed dashboard and the parameters are changing at the same time', async () => {
    const { dashboardFrame, component } = setupComponent()
    await nextTick() // wait until first setup is finished
    await component.setProps({
      dashboard: 'another-dashboard',
      parameters: contentOptions.parameters
    })
    expect(dashboardFrame.navigateToDashboard).toHaveBeenCalledOnce()
    expect(dashboardFrame.navigateToDashboard).toHaveBeenCalledWith('another-dashboard', {
      parameters: contentOptions.parameters
    })
  })
})
