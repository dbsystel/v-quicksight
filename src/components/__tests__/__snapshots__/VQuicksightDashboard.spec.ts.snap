// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VQuicksightDashboard > renders as expected 1`] = `
<div
  class="v-quicksight-dashboard"
  id="v-quicksight-dashboard-1"
/>
`;
